Bottom: 8d57d68463cc91b126eea23fd6546e1cd233b85c
Top:    58424267ac79a08264f4a2fdac7780d19de5ad93
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2016-01-17 13:09:57 +0000

drc: Fix polygon-polygon clearance checks

IsPolygonInPolygon was applying 2x bloat to the line it uses to test with
against the second polygon, and IsLineInPolygon was applying a further
bloat offset - making the clearance used incorrect.

This caused false DRC violation reports of polygons being too close to
one-another, despite there being adequate clearance.

Note for future - a cleaner fix regarding "Bloat", and the DRC, is that
we should probably pass it more explicitly as a parameter where it is
to be used, in order to avoid surprises as to where it gets applied.


---

diff --git a/src/find.c b/src/find.c
index e1c1ee2..92f9d6d 100644
--- a/src/find.c
+++ b/src/find.c
@@ -2504,7 +2504,12 @@ IsPolygonInPolygon (PolygonType *P1, PolygonType *P2)
 
               line.Point1.X = v->point[0];
               line.Point1.Y = v->point[1];
-              line.Thickness = 2 * Bloat;
+              line.Thickness = Bloat;
+              /* Another Bloat is added by IsLineInPolygon, making the correct
+               * 2x Bloat. Perhaps we should change it there, but doing so
+               * breaks some other DRC checks which rely on the behaviour
+               * in IsLineInPolygon.
+               */
               line.Clearance = 0;
               line.Flags = NoFlags ();
               for (v = v->next; v != &c->head; v = v->next)
